### ISSUE TITLE
NavigationEnd resolves before DOM is updated on initial load with nested router-outlets

### ISSUE THREAD

**Issue by bjarketrux** [angular/angular#59065]
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

No

### Description

In our application we follow the accessibility pattern: [Focus management after navigation](https://angular.dev/best-practices/a11y#focus-management-after-navigation).

This has worked for a while, but recently we introduced an extra LayoutComponent in the router turning the existing components into children of it. This caused a bug as the element to focus on `NavigationEnd` does not get focus on the initial page load (disregard that this might be better for accessibility).

In the attached StackBlitz I have recreated the "bug" (or inconsistency). There is a `debugger;` statement at the line where the focus is supposed to be set (open DevTools to make it trigger).

### Scenario 1 (not as expected)
1. Open DevTools on the StackBlitz example
2. Click the preview reload icon to make the preview page reload
3. **Observe**: When the break point is hit, the `Layout2` component and the `Focus` div is not rendered in the DOM (but `Main` and `Layout1` is)

### Scenario 2 (as expected)
1. Continue from previous example
2. Click continue to continue execution
3. Press the Page 1 or Page 2 link
4. **Observe**: When the break point is hit, the `Layout2` component and the `Focus` div **is** rendered in the DOM

### Expected behavior
I expected (maybe wrongly) that when `NavigationEnd` is triggered, all router-outlets have been rendered to DOM.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-yue3ta?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true

```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 19.0.0
Node: 18.20.3
Package Manager: npm 10.2.3
OS: linux x64

Angular: 19.0.0
... animations, build, cli, common, compiler, compiler-cli, core
... forms, platform-browser, router

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.1900.0
@angular-devkit/core         19.0.0
@angular-devkit/schematics   19.0.0
@schematics/angular          19.0.0
rxjs                         7.8.1
typescript                   5.6.3
zone.js                      0.15.0
```

### Anything else?

A possible workaround - if the navigation event works as expected - is to add a `setTimeout` around the DOM manipulation (maybe?).

**Comment by fyllekanin** [comment]
Not sure if exact same issue but similar I imagine (flicker in the sense that initial / renders first then the router kicks in and render the actual nested routing).

Small workaround I got working for it:
```
private readonly isServer: boolean = isPlatformServer(inject(PLATFORM_ID));

ngOnInit(): void {
    if (this.isServer) {
        return;
    }
    this.router.events.subscribe(event => {
        if (event instanceof ChildActivationEnd) {
            // Nested routes are finished
            // Example update a variable to true saying it's finished which controls a display property
            // to show the main router outlet or in this case trigger the scroll
        }
    });
}

```

**Comment by atscott** [comment]
Yes, this is expected. The Router's events, including `NavigationEnd` have to do with the lifecycle of the routing logic and are not necessarily related to Angular's rendering. If you want to wait for the next rendering cycle, you can use `afterNextRender` in your `NavigationEnd` subscription. 

I do agree that this is somewhat unfortunate. Prior to the introduction of the `afterNextRender` API, Angular did not have a way to hook into the global rendering, so the Router did not have the ability to effectively wait for it to happen before firing `NavigationEnd`. The timings of the events in the Router are more or less cemented where they are. Changing them causes unexpected breakages in applications so the current timing of `NavigationEnd` will remain unless there's a _really_ good reason that justifies the churn.

**Comment by bjarketrux** [comment]
# Solution
The following snippet is what we ended up having.

**Note**: You need to use `runInInjectionContext` to use `afterNextRender`, see #55870.

```typescript
constructor(router: Router, viewportScroller: ViewportScroller, environmentInjector: EnvironmentInjector) {
    router.events.pipe(takeUntilDestroyed()).subscribe({
      next: (routerEvent) => {
        if (routerEvent instanceof NavigationEnd) {
          // Accessibility focus management: https://angular.dev/best-practices/a11y#focus-management-after-navigation
          runInInjectionContext(environmentInjector, () =>
            afterNextRender({ write: () => viewportScroller.scrollToAnchor('main-content-focus') }),
          );
        }
      },
    });
  }
```

StackBlitz for the solution: https://stackblitz.com/edit/stackblitz-starters-mcpy7dcg?file=src%2Fmain.ts

**Comment by JeanMeche** [comment]
Fwiw, you can pass the injector directly: 

```ts
afterNextRender(
  {
    write: () => {
      debugger;
      viewportScroller.scrollToAnchor('main-content-focus');
    },
  },
  { injector: environmentInjector }
);
```
